Handle Opencast starting up before index is available

gregorydlogan opened the pull request:
This PR supercedes #5944, handling additional cases which caused the initial PR to hang or otherwise misbehave.

Fixes #5943

### How to test this patch

Make sure Opensearch is not running, then start Opencast an observe the waiting loop.  Play with the timing.  Then start Opensearch.

You may see exceptions related to circular dependencies.  These happen without that patch, and are their own separate issue.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] ~include migration scripts and documentation, if appropriate~
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [x] ~explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch~

marwyg committed:
Fix Opencast startup failure when opensearch boots later

Adds a check in the AbstractElasticsearchIndex.init() method. The check incudes sending
a request to opensearch and evaluating the result. If the service
is not reachable, a retry loop will continuously check the connection.

gregorydlogan committed:
Testing not just that we can connect to elasticsearch, but that the health status is good. Also handling the case where we catch Elastic mid-boot, which generates a ConnectionClosedException which casts to a RuntimeException which was allowing things to start starting up prior to ES/OS being ready for it.

KatrinIhler commented:
Soo, technically this is a bug fix, so I'd say this can go into 17. Of course I have a vested interest in seeing this go into 16 as well because then we don't have to work around this issue for those systems. @lkiesow Do you have an opinion on this?

lkiesow commented:
I don't care

marwyg reviewed (approved):
The changes in this PR were already approved in PR #5944, with one change request: to use the original connection method instead of the one I had used.

This PR addresses that change request by switching out the connection method. I tested it in my local setup with OpenSearch, and everything worked as expected. Additionally, I tested changing the "retry delay" to a small value (10ms), and it worked fine.

The code changes also look good to me.

Thank you @gregorydlogan! This PR looks good to me and can be merged, as far as I am concerned.

marwyg commented:
I also tested this with Branch 16.x right now. It works as expected.

marwyg commented:
> The changes in this PR were already approved in PR #5944, with one change request: to use the original connection method instead of the one I had used.
> 
> This PR addresses that change request by switching out the connection method. I tested it in my local setup with OpenSearch, and everything worked as expected. Additionally, I tested changing the "retry delay" to a small value (10ms), and it worked fine.
> 
> The code changes also look good to me.
> 
> Thank you @gregorydlogan! This PR looks good to me and can be merged, as far as I am concerned.

Ah okay.. I mixed the PRs up. The problem was not the connection method. It was that if starting ES/OS via container, the module wouldn't start properly. I tested this with a container and got the same error, I will investigate this.

marwyg committed:
Make starting containers also work

Previously the retry mechanism would work in case opensearch would start
as a service (and of course after opencast). With this change it also
works with opensearch in a container. In case a container is started,
an IOException will be thrown with a SocketException as the cause, so this
commit adds a catch for this case and handles it.

marwyg commented:
I created a PR against your branch @gregorydlogan. I added another catch that deals with the case, when opensearch is started in a container.

gregorydlogan commented:
As discussed in the dev meeting, this would be nice to have in 16.x.  I will rebase this to 16.x and then we can merge it.

gregorydlogan commented:
@KatrinIhler Done, you wanna merge this since both Martin and I have touched it?

KatrinIhler committed:
Update elasticsearch.md

Rename "index service" to "index"

KatrinIhler committed:
Update org.opencastproject.elasticsearch.index.ElasticsearchIndex.cfg

Renamed "index service" to "index"

KatrinIhler commented:
Changed "index service" to "index" to avoid confusion with Opencast's internal index service. Won't nitpick at the code bc of time, overall execution looks good.
